packages/editor/src/extensions/heading/heading.ts

```diff
@@ -71,5 +100,140 @@ export const Heading = TiptapHeading.extend({
+                  const nextNode = doc.nodeAt(nextPos);
+                  if (!nextNode) break;
+                  if (
+                    nextNode.type.name === "heading" &&
+                    nextNode.attrs.level <= headingLevel
+                  ) {
+                    break;
+                  }
+
+                  decorations.push(
+                    Decoration.node(nextPos, nextPos + nextNode.nodeSize, {
+                      style: "display: none;"
+                    })
+                  );
+
+                  nextPos += nextNode.nodeSize;
+                }
+              }
+            });
+
+            return DecorationSet.create(doc, decorations);
+          }
+        }
+      })
+    ];
+  },
+
+  addNodeView() {
+    return ({ node, getPos, editor, HTMLAttributes }) => {
+      const heading = document.createElement(`h${node.attrs.level}`);
+
+      for (const attr in HTMLAttributes) {
+        heading.setAttribute(attr, HTMLAttributes[attr]);
+      }
+
+      if (node.attrs.collapsed) heading.dataset.collapsed = "true";
+      else delete heading.dataset.collapsed;
+
+      function onClick(e: MouseEvent | TouchEvent) {
+        if (e instanceof MouseEvent && e.button !== 0) return;
+        if (!(e.target instanceof HTMLHeadingElement)) return;
+
+        const pos = typeof getPos === "function" ? getPos() : 0;
+        if (typeof pos !== "number") return;
+
+        const resolvedPos = editor.state.doc.resolve(pos);
+        const calloutAncestor = findParentNodeClosestToPos(
+          resolvedPos,
+          (node) => node.type.name === "callout"
+        );
+        if (calloutAncestor) return;
+
+        const { x, y, right } = heading.getBoundingClientRect();
+
+        const clientX =
+          e instanceof MouseEvent ? e.clientX : e.touches[0].clientX;
+        const clientY =
+          e instanceof MouseEvent ? e.clientY : e.touches[0].clientY;
+
+        const hitArea = { width: 40, height: 40 };
+
+        const isRtl =
+          heading.dir === "rtl" ||
+          findParentNodeClosestToPos(
+            resolvedPos,
+            (node) => !!node.attrs.textDirection
+          )?.node.attrs.textDirection === "rtl";
+
+        let xStart = clientX >= x - hitArea.width;
+        let xEnd = clientX <= x;
+        const yStart = clientY >= y;
+        const yEnd = clientY <= y + hitArea.height;
+
+        if (isRtl) {
+          xEnd = clientX <= x + hitArea.width;
+          xStart = clientX >= right;
+        }
+
+        if (xStart && xEnd && yStart && yEnd) {
+          e.preventDefault();
+          e.stopImmediatePropagation();
+
+          editor.commands.command(({ tr }) => {
+            const currentNode = tr.doc.nodeAt(pos);
+            if (currentNode) {
+              tr.setNodeAttribute(
+                pos,
+                "collapsed",
+                !currentNode.attrs.collapsed
+              );
+            }
+            return true;
+          });
+        }
+      }
```

Comment:
This is the same function as used in Outline right? Maybe we can share the logic?

Reply:
I tried to make the entire onClick reusable, but I felt we were passing in a lot of parameters and conditional logic making the function difficult to understand. I've made a util function for getting the click bounds.